docs/gate-configuration.md
# NeoFS REST Gateway configuration file

This section contains detailed NeoFS REST Gateway configuration file description
including default config values and some tips to set up configurable values.

There are some custom types used for brevity:

* `duration` -- string consisting of a number and a suffix. Suffix examples include `s` (seconds), `m` (minutes), `ms` (
  milliseconds).

# Structure

| Section      | Description                                     |
|--------------|-------------------------------------------------|
| `server`     | [Server configuration](#server-section)         |
| `wallet`     | [Wallet configuration](#wallet-section)         |
| `pool`       | [Pool configuration](#pool-section)             |
| `logger`     | [Logger configuration](#logger-section)         |
| `pprof`      | [Pprof configuration](#pprof-section)           |
| `prometheus` | [Prometheus configuration](#prometheus-section) |

# `server` section

```yaml
server:
  endpoints:
    - address: localhost:8080
      external-address: localhost:8090
      tls:
        enabled: false
        certificate: /path/to/tls/cert
        key: /path/to/tls/key
        ca-certificate: /path/to/tls/ca
      keep-alive: 3m
      read-timeout: 60s
      write-timeout: 60s
  cleanup-timeout: 10s
  graceful-timeout: 15s
  max-header-size: 1000000
  listen-limit: 0
```

| Parameter                       | Type       | Default value    | Description                                                                                                                                                                         |
|---------------------------------|------------|------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `cleanup-timeout`               | `duration` | `10s`            | Grace period for which to wait before killing idle connections.                                                                                                                     |
| `graceful-timeout`              | `duration` | `15s`            | Grace period for which to wait before shutting down the server.                                                                                                                     |
| `max-header-size`               | `int`      | `1000000`        | Controls the maximum number of bytes the server will read parsing the request header's keys and values, including the request line. It does not limit the size of the request body. |
| `listen-limit`                  | `int`      | `0`              | Limit the number of outstanding requests. `0` means no limit                                                                                                                        |                                                                                                                         |
| `endpoint.[0].address`          | `string`   | `localhost:8080` | The IP and port to listen on.                                                                                                                                                       |
| `endpoint.[0].keep-alive`       | `duration` | `3m`             | Sets the TCP keep-alive timeouts on accepted connections.                                                                                                                           |
| `endpoint.[0].read-timeout`     | `duration` | `60s`            | Maximum duration before timing out read of the request. It prunes dead TCP connections (e.g. closing laptop mid-download).                                                          |
| `endpoint.[0].write-timeout`    | `duration` | `60s`            | Maximum duration before timing out write of the response.                                                                                                                           |
| `endpoint.[0].tls.enabled`      | `bool`     | `false`          | Use TLS for a gRPC connection (min version is TLS 1.2).                                                                                                                             |
| `endpoint.[0].tls.certificate`  | `string`   |                  | The certificate file to use for secure connections.                                                                                                                                 |
| `endpoint.[0].tls.key`          | `string`   |                  | The private key file to use for secure connections (without passphrase).                                                                                                            |
| `endpoint.[0].tls.ca`           | `string`   |                  | The certificate authority certificate file to be used with mutual tls auth.                                                                                                         |
| `endpoint.[0].external-address` | `string`   |                  | The full URL address needs to be shown in the API documentation, including the scheme (http/https), host, and port. If not set, will be generated from `address` and `tls.enabled`. |

# `wallet` section

```yaml
wallet:
  path: /path/to/wallet.json
  address: NfgHwwTi3wHAS8aFAN243C5vGbkYDpqLHP
  passphrase: pwd
```

| Parameter    | Type     | Default value | Description                                                              |
|--------------|----------|---------------|--------------------------------------------------------------------------|
| `path`       | `string` |               | Path to the wallet.                                                      |
| `address`    | `string` |               | Account address to get from wallet. If omitted default one will be used. |
| `passphrase` | `string` |               | Passphrase to decrypt wallet.                                            |

# `pool` section

```yaml
pool:
  node-dial-timeout: 10s
  healthcheck-timeout: 15s
  rebalance-timer: 60s
  error-threshold: 100

  # Nodes configuration
  # This configuration makes the gateway use the first node (node1.neofs:8080)
  # while it's healthy. Otherwise, gateway uses the second node (node2.neofs:8080)
  # for 10% of requests and the third node (node3.neofs:8080) for 90% of requests.
  # Until nodes with the same priority level are healthy
  # nodes with other priority are not used.
  # The lower the value, the higher the priority.
  peers:
    - address: node1.neofs:8080
      priority: 1
      weight: 1
    - address: node2.neofs:8080
      priority: 2
      weight: 0.1
    - address: node3.neofs:8080
      priority: 2
      weight: 0.9
```

| Parameter               | Type       | Default value | Description                                                                                             |
|-------------------------|------------|---------------|---------------------------------------------------------------------------------------------------------|
| `node-dial-timeout`     | `duration` | `10s`         | Timeout to connect to a node.                                                                           |
| `healthcheck-timeout`   | `duration` | `15s`         | Timeout to check node health during rebalance.                                                          |
| `rebalance-timer`       | `duration` | `60s`         | Interval to check node health.                                                                          |
| `error-threshold`       | `uint32`   | `100`         | The number of errors on connection after which node is considered as unhealthy.                         |
| `default-timestamp`     | `bool`     | `false`       | By default, add a Timestamp to the object upon creation.                                                |

## `peers` section

| Parameter              | Type       | Default value | Description                                                                                                                                             |
|------------------------|------------|---------------|---------------------------------------------------------------------------------------------------------------------------------------------------------|
| `address`              | `string`   |               | Address of storage node.                                                                                                                                |
| `priority`             | `int`      | `1`           | It allows to group nodes and don't switch group until all nodes with the same priority will be unhealthy. The lower the value, the higher the priority. |
| `weight`               | `float`    | `1`           | Weight of node in the group with the same priority. Distribute requests to nodes proportionally to these values.                                        |


CHANGELOG.md
# Changelog

This document outlines major changes between releases.

## [Unreleased]

### Added

### Fixed

### Changed
- Go 1.25+ is required to build now (#322)

### Removed
- `pool.container-ops-poll-interval` option (#361)
- `pool.container-ops-timeoutl` option (#361)

### Updated

### Upgrading from 0.16.0

## [0.16.0] - 2026-02-11

### Added
- Session token v2 support with new auth/token model (#341, #344, #351, #353, #355)
- Gateway metadata (#341)
- Support for numeric EACL filters (#347)
- Support for account-based EACL targets (#350)

### Fixed
- Excessive CORS headers pushed to users via OPTIONS (#354)

### Removed
- Support for session token v1 in container operations (#351)
- Redundant KEYS role for EACL targets (#350)
- Unused data types (#354)

### Updated
- NeoFS SDK dependency to RC17 (#339, #341, #351)

### Upgrading from 0.15.1
Applications using REST gateway for container operations must migrate to the new
authentication scheme immediately, this version in incompatible with container
operation tokens used by 0.15.1. Migration to new tokens for object operations is
also recommended since old authentication scheme is deprecated and eventually
will be deleted. See [docs/migration-to-auth-v2.md](./docs/migration-to-auth-v2.md)
for migration details.

## [0.15.1] - 2025-11-21

### Updated
- NeoGo dependency to 0.114.0 (#334)
- NeoFS SDK dependency to RC16+ fixing Incomplete status processing (#334)
- golang.org/x/crypto dependency from 0.42.0 to 0.45.0 (#335)

## [0.15.0] - 2025-11-13

### Added
- Incomplete flag to `/v2/objects/{cid}/search` operation (#329)

### Fixed
- "Incomplete" code handling for implicit searches (#331)

### Removed
- Deprecated `/v1/objects`, `/v1/objects/{containerId}/{objectId}`, `/v1/get/{containerId}/{objectId}`, `/v1/upload/{containerId}` and `/v1/get_by_attribute/{containerId}/{attrKey}/{attrVal}` APIs (#330)

### Upgrading from 0.14.2
Please refer to migration guides for [new upload](./docs/migration-new-upload.md)
and [HTTP](./docs/migration-from-http.md) in case your application still uses
any of the removed APIs.

## [0.14.2] - 2025-10-15

### Fixed
- Support for ROOT and PHY filters (#326)

### Changed
- Go 1.24+ is required to build now (#266)

### Updated
- github.com/nspcc-dev/neo-go v0.111.0 => v0.113.0 (#266, #325)
- github.com/spf13/pflag v1.0.5 => v1.0.10 (#266)
- github.com/stretchr/testify v1.10.0 => v1.11.1 (#266)
- NeoFS SDK to RC15 (#325)

## [0.14.1] - 2025-08-12

### Fixed
- Panic in auth requests without X-Bearer-For-All-Users parameter (#320)

### Updated
- NeoFS SDK to RC14 version (#319)
- github.com/nspcc-dev/neo-go v0.108.1 => v0.111.0 (#318)
- github.com/labstack/echo/v4 v4.13.3 => v4.13.4 (#318)

## [0.14.0] - 2025-07-30

### Added
- `pool.container-ops-poll-interval` option to control polling behavior for container operations (#315)

### Fixed
- Expiration parameter using pre-2.18 API meaning of EpochDuration (blocks vs seconds, #313)
- Incorrect error code for wrong EACL tables (#315)

### Changed
- Default container-ops-timeout to 10s (previously dynamic, based on block time, #313)
- Network info is cached for half epoch duration now (#313)
- Default polling interval for container operations from fixed 1s to dynamic value based on block time (#315)

## [0.13.0] - 2025-07-22

### Fixed
- Panic on search v1 (#305, #308)

### Changed
- Peers list definition in config files (#302)
- `pool.container_ops_timeout` name to `pool.container-ops-timeout` (#307)

### Updated
- NeoFS SDK to v1.0.0-rc.13.0.20250715070617-c7038b450691 (API 2.18, #311)

### Upgrading from 0.12.0
Rename `pool.container_ops_timeout` option to `pool.container-ops-timeout`
if used. Update peer configuration from the old format:
```yaml
peers:
  0:
    address: node1.neofs:8080
    priority: 1
    weight: 1
```
to the new one:
```yaml
peers:
  - address: node1.neofs:8080
    priority: 1
    weight: 1
```

## [0.12.0] - 2025-04-30

### Added
- `logger.encoding` config option (#273)
- `logger.timestamp` config option (#273)
- `/v2/objects/{containerId}/search` entrypoint with `cursor` support (instead of `offset`) for objects search (#276, #294)
- Ability to request and return attributes for `search v2` (#277)
- `pool.container_ops_timeout` config option (#254)
- POST /containers API replacing the old PUT-based one (#287)
- Request processing histograms in metrics (#285)
- `MatchNumGT`, `MatchNumGE`, `MatchNumLT`, `MatchNumLE` operators to search (#260)


cmd/neofs-rest-gw/config.go
package main

import (
	"context"
	"fmt"
	"os"
	"runtime"
	"slices"
	"strconv"
	"strings"
	"time"

	"github.com/nspcc-dev/neo-go/cli/flags"
	"github.com/nspcc-dev/neo-go/cli/input"
	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
	"github.com/nspcc-dev/neo-go/pkg/util"
	"github.com/nspcc-dev/neo-go/pkg/wallet"
	"github.com/nspcc-dev/neofs-rest-gw/handlers"
	"github.com/nspcc-dev/neofs-rest-gw/metrics"
	"github.com/nspcc-dev/neofs-sdk-go/client"
	"github.com/nspcc-dev/neofs-sdk-go/pool"
	"github.com/nspcc-dev/neofs-sdk-go/user"
	"github.com/spf13/pflag"
	"github.com/spf13/viper"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
	"golang.org/x/term"
)

const (
	defaultConnectTimeout     = 10 * time.Second
	defaultHealthcheckTimeout = 15 * time.Second
	defaultRebalanceTimer     = 60 * time.Second

	defaultShutdownTimeout = 15 * time.Second

	defaultPoolErrorThreshold   uint32 = 100
	defaultPoolDefaultTimestamp bool   = false

	// default [handlers.PrmAPI.MaxPayloadBufferSize] value.
	defaultMaxObjectPayloadBufferSize = 4 << 20

	// Pool config.
	cmdNodeDialTimeout      = "node-dial-timeout"
	cfgNodeDialTimeout      = "pool." + cmdNodeDialTimeout
	cmdHealthcheckTimeout   = "healthcheck-timeout"
	cfgHealthcheckTimeout   = "pool." + cmdHealthcheckTimeout
	cmdRebalance            = "rebalance-timer"
	cfgRebalance            = "pool." + cmdRebalance
	cfgPoolErrorThreshold   = "pool.error-threshold"
	cfgPoolDefaultTimestamp = "pool.default-timestamp"
	cmdPeers                = "peers"
	cfgPeers                = "pool." + cmdPeers
	cfgPeerAddress          = "address"
	cfgPeerPriority         = "priority"
	cfgPeerWeight           = "weight"

	// Metrics / Profiler.
	cfgPrometheusEnabled = "prometheus.enabled"
	cfgPrometheusAddress = "prometheus.address"
	cfgPprofEnabled      = "pprof.enabled"
	cfgPprofAddress      = "pprof.address"

	// Logger.
	cfgLoggerLevel     = "logger.level"
	cfgLoggerEncoding  = "logger.encoding"
	cfgLoggerTimestamp = "logger.timestamp"

	// Wallet.
	cfgWalletPath       = "wallet.path"
	cfgWalletAddress    = "wallet.address"
	cfgWalletPassphrase = "wallet.passphrase"

	// Server endpoints.
	cfgServerSection   = "server."
	cfgServerEndpoints = cfgServerSection + "endpoints"
	cfgServerNNSName   = cfgServerSection + "nns-name"

	cfgTLSEnabled    = "tls.enabled"
	cfgTLSKeyFile    = "tls.key"
	cfgTLSCertFile   = "tls.certificate"
	cfgTLSCertCAFile = "tls.ca-certificate"

	cfgEndpointAddress         = "address"
	cfgEndpointExternalAddress = "external-address"
	cfgEndpointKeepAlive       = "keep-alive"
	cfgEndpointReadTimeout     = "read-timeout"
	cfgEndpointWriteTimeout    = "write-timeout"

	// Command line args.
	cmdHelp          = "help"
	cmdVersion       = "version"
	cmdPprof         = "pprof"
	cmdMetrics       = "metrics"
	cmdWallet        = "wallet"
	cmdAddress       = "address"
	cmdConfig        = "config"
	cmdListenAddress = "listen-address"

	baseURL = ""
)

var ignore = map[string]struct{}{
	cmdPeers:   {},
	cmdHelp:    {},
	cmdVersion: {},
}

// Prefix is a prefix used for environment variables containing gateway
// configuration.
const Prefix = "REST_GW"

var (
	// Version is gateway version.
	Version = "dev"
)

func config() *viper.Viper {
	v := viper.New()
	v.AutomaticEnv()
	v.SetEnvPrefix(Prefix)
	v.AllowEmptyEnv(true)
	v.SetConfigType("yaml")
	v.SetEnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_"))

	// flags setup:
	flagSet := pflag.NewFlagSet("commandline", pflag.ExitOnError)
	flagSet.SetOutput(os.Stdout)
	flagSet.SortFlags = false

	flagSet.Bool(cmdPprof, false, "enable pprof")
	flagSet.Bool(cmdMetrics, false, "enable prometheus")

	help := flagSet.BoolP(cmdHelp, "h", false, "show help")
	version := flagSet.BoolP(cmdVersion, "v", false, "show version")

	flagSet.StringP(cmdWallet, "w", "", `path to the wallet`)
	flagSet.String(cmdAddress, "", `address of wallet account`)

	configFlag := flagSet.String(cmdConfig, "", "config path")
	flagSet.Duration(cmdNodeDialTimeout, defaultConnectTimeout, "gRPC node connect timeout")
	flagSet.Duration(cmdHealthcheckTimeout, defaultHealthcheckTimeout, "gRPC healthcheck timeout")
	flagSet.Duration(cmdRebalance, defaultRebalanceTimer, "gRPC connection rebalance timer")

	peers := flagSet.StringArrayP(cmdPeers, "p", nil, "NeoFS nodes")

	flagSet.String(cmdListenAddress, "localhost:8080", "set the main address to listen")
	flagSet.String(cfgTLSCertFile, "", "TLS certificate file to use; note that if you want to start HTTPS server, you should also set up --"+cmdListenAddress+" and --"+cfgTLSKeyFile)
	flagSet.String(cfgTLSKeyFile, "", "TLS key file to use; note that if you want to start HTTPS server, you should also set up --"+cmdListenAddress+" and --"+cfgTLSCertFile)
	flagSet.Duration(cfgEndpointKeepAlive, 3*time.Minute, "sets the TCP keep-alive timeouts on accepted connections. It prunes dead TCP connections ( e.g. closing laptop mid-download)")
	flagSet.Duration(cfgEndpointReadTimeout, 60*time.Second, "maximum duration before timing out read of the request")
	flagSet.Duration(cfgEndpointWriteTimeout, 60*time.Second, "maximum duration before timing out write of the response")
	flagSet.String(cfgEndpointExternalAddress, "", "the full URL address needs to be shown in the API documentation")

	// init server flags
	BindDefaultFlags(flagSet)

	if err := bindServerFlags(v, flagSet); err != nil {
		panic(fmt.Errorf("bind server flags: %w", err))
	}
	// set defaults:
	// pool
	v.SetDefault(cfgPoolErrorThreshold, defaultPoolErrorThreshold)
	v.SetDefault(cfgPoolDefaultTimestamp, defaultPoolDefaultTimestamp)

	// metrics
	v.SetDefault(cfgPprofAddress, "localhost:8091")
	v.SetDefault(cfgPrometheusAddress, "localhost:8092")

	// logger:
	v.SetDefault(cfgLoggerLevel, "debug")
	v.SetDefault(cfgLoggerEncoding, "console")

	// Bind flags
	for cfg, cmd := range bindings {
		if err := v.BindPFlag(cfg, flagSet.Lookup(cmd)); err != nil {
			panic(fmt.Errorf("bind flags: %w", err))
		}
	}

	if err := flagSet.Parse(os.Args); err != nil {
		panic(err)
	}

	switch {
	case help != nil && *help:
		fmt.Printf("NeoFS REST Gateway %s\n", Version)
		flagSet.PrintDefaults()

		fmt.Println()
		fmt.Println("Default environments:")
		fmt.Println()
		cmdKeys := v.AllKeys()
		slices.Sort(cmdKeys)

		for i := range cmdKeys {
			if _, ok := ignore[cmdKeys[i]]; ok {
				continue
			}

			k := strings.ReplaceAll(cmdKeys[i], ".", "_")
			fmt.Printf("%s_%s = %v\n", Prefix, strings.ToUpper(k), v.Get(cmdKeys[i]))
		}

		os.Exit(0)
	case version != nil && *version:
		fmt.Printf("NeoFS REST Gateway\nVersion: %s\nGoVersion: %s\n", Version, runtime.Version())
		os.Exit(0)
	case configFlag != nil && *configFlag != "":
		if cfgFile, err := os.Open(*configFlag); err != nil {
			panic(err)
		} else if err = v.ReadConfig(cfgFile); err != nil {
			panic(err)
		}
	}

	if peers != nil && len(*peers) > 0 {
		for i := range *peers {
			v.SetDefault(cfgPeers+"."+strconv.Itoa(i)+"."+cfgPeerAddress, (*peers)[i])
			v.SetDefault(cfgPeers+"."+strconv.Itoa(i)+"."+cfgPeerWeight, 1)
			v.SetDefault(cfgPeers+"."+strconv.Itoa(i)+"."+cfgPeerWeight, 1)
		}
	}

	return v
}

func bindServerFlags(v *viper.Viper, flags *pflag.FlagSet) error {
	// This key is used only to check if the address comes from the command arguments.
	if err := v.BindPFlag(cmdListenAddress, flags.Lookup(cmdListenAddress)); err != nil {
		return err
	}

	if err := v.BindPFlag(cfgServerEndpoints+".0."+cfgEndpointAddress, flags.Lookup(cmdListenAddress)); err != nil {
		return err
	}
	if err := v.BindPFlag(cfgServerEndpoints+".0."+cfgEndpointExternalAddress, flags.Lookup(cfgEndpointExternalAddress)); err != nil {
		return err
	}
	if err := v.BindPFlag(cfgServerEndpoints+".0."+cfgEndpointKeepAlive, flags.Lookup(cfgEndpointKeepAlive)); err != nil {
		return err
	}
	if err := v.BindPFlag(cfgServerEndpoints+".0."+cfgEndpointReadTimeout, flags.Lookup(cfgEndpointReadTimeout)); err != nil {
		return err
	}
	if err := v.BindPFlag(cfgServerEndpoints+".0."+cfgEndpointWriteTimeout, flags.Lookup(cfgEndpointWriteTimeout)); err != nil {
		return err
	}
	if err := v.BindPFlag(cfgServerEndpoints+".0."+cfgTLSKeyFile, flags.Lookup(cfgTLSKeyFile)); err != nil {
		return err
	}

	return v.BindPFlag(cfgServerEndpoints+".0."+cfgTLSCertFile, flags.Lookup(cfgTLSCertFile))
}

func init() {
	for _, flagName := range serverFlags {
		cfgName := cfgServerSection + flagName
		bindings[cfgName] = flagName
		knownConfigParams[cfgName] = struct{}{}
	}
}

var serverFlags = []string{
	FlagCleanupTimeout,
	FlagGracefulTimeout,
	FlagMaxHeaderSize,
	FlagListenLimit,
}

var bindings = map[string]string{
	cfgPprofEnabled:       cmdPprof,
	cfgPrometheusEnabled:  cmdMetrics,
	cfgNodeDialTimeout:    cmdNodeDialTimeout,
	cfgHealthcheckTimeout: cmdHealthcheckTimeout,
	cfgRebalance:          cmdRebalance,
	cfgWalletPath:         cmdWallet,
	cfgWalletAddress:      cmdAddress,
}

var knownConfigParams = map[string]struct{}{
	cfgWalletAddress:        {},
	cfgWalletPath:           {},
	cfgWalletPassphrase:     {},
	cfgRebalance:            {},
	cfgHealthcheckTimeout:   {},
	cfgNodeDialTimeout:      {},
	cfgPoolErrorThreshold:   {},
	cfgPoolDefaultTimestamp: {},
	cfgLoggerLevel:          {},
	cfgLoggerEncoding:       {},
	cfgLoggerTimestamp:      {},
	cfgPrometheusEnabled:    {},
	cfgPrometheusAddress:    {},
	cfgPprofEnabled:         {},
	cfgPprofAddress:         {},
	cmdListenAddress:        {},
}

func validateConfig(cfg *viper.Viper, logger *zap.Logger) {
	peerNumsMap := make(map[int]struct{})

	for _, providedKey := range cfg.AllKeys() {
		if !strings.HasPrefix(providedKey, cfgPeers) {
			if strings.HasPrefix(providedKey, cfgServerEndpoints) {
				// Do not validate `Endpoints` section.
				continue
			}
			if _, ok := knownConfigParams[providedKey]; !ok {
				logger.Warn("unknown config parameter", zap.String("key", providedKey))
			}
			continue
		}

		num, ok := isValidPeerKey(providedKey)
		if !ok {
			logger.Warn("unknown config parameter", zap.String("key", providedKey))
		} else {
			peerNumsMap[num] = struct{}{}
		}
	}

	peerNums := make([]int, 0, len(peerNumsMap))
	for num := range peerNumsMap {
		peerNums = append(peerNums, num)
	}
	slices.Sort(peerNums)

	for i, num := range peerNums {
		if i != num {
			logger.Warn("invalid config parameter, peer indexes must be consecutive starting from 0", zap.String("key", cfgPeers+"."+strconv.Itoa(num)))
		}
	}
}

func isValidPeerKey(key string) (int, bool) {
	trimmed := strings.TrimPrefix(key, cfgPeers)
	split := strings.Split(trimmed, ".")

	if len(split) != 3 {
		return 0, false
	}

	if split[2] != cfgPeerAddress && split[2] != cfgPeerPriority && split[2] != cfgPeerWeight {
		return 0, false
	}

	num, err := strconv.Atoi(split[1])
	if err != nil || num < 0 {
		return 0, false
	}

	return num, true
}

func getNeoFSKey(logger *zap.Logger, cfg *viper.Viper) (*keys.PrivateKey, error) {
	walletPath := cfg.GetString(cmdWallet)
	if len(walletPath) == 0 {
		walletPath = cfg.GetString(cfgWalletPath)
	}

	if len(walletPath) == 0 {
		logger.Info("no wallet path specified, creating ephemeral key automatically for this run")
		return keys.NewPrivateKey()
	}
	w, err := wallet.NewWalletFromFile(walletPath)
	if err != nil {
		return nil, err
	}

	var password *string
	if cfg.IsSet(cfgWalletPassphrase) {
		pwd := cfg.GetString(cfgWalletPassphrase)
		password = &pwd
	}

	address := cfg.GetString(cmdAddress)
	if len(address) == 0 {
		address = cfg.GetString(cfgWalletAddress)
	}

	return getKeyFromWallet(w, address, password)
}

func getKeyFromWallet(w *wallet.Wallet, addrStr string, password *string) (*keys.PrivateKey, error) {
	var addr util.Uint160
	var err error

	if addrStr == "" {
		addr = w.GetChangeAddress()
	} else {
		addr, err = flags.ParseAddress(addrStr)
		if err != nil {
			return nil, fmt.Errorf("invalid address: %w", err)
		}
	}

	acc := w.GetAccount(addr)
	if acc == nil {
		return nil, fmt.Errorf("couldn't find wallet account for %s", addrStr)
	}

	if password == nil {
		pwd, err := input.ReadPassword("Enter password > ")
		if err != nil {
			return nil, fmt.Errorf("couldn't read password: %w", err)
		}
		password = &pwd
	}

	if err := acc.Decrypt(*password, w.Scrypt); err != nil {
		return nil, fmt.Errorf("couldn't decrypt account: %w", err)
	}

	return acc.PrivateKey(), nil
}

// newLogger constructs a zap.Logger instance for current application.
// Panics on failure.
//
// Logger is built from zap's production logging configuration with:
//   - parameterized level (debug by default)
//   - console encoding
//   - ISO8601 time encoding
//
// Logger records a stack trace for all messages at or above fatal level.
//
// See also zapcore.Level, zap.NewProductionConfig, zap.AddStacktrace.
func newLogger(v *viper.Viper) *zap.Logger {
	var lvl zapcore.Level
	lvlStr := v.GetString(cfgLoggerLevel)
	err := lvl.UnmarshalText([]byte(lvlStr))
	if err != nil {
		panic(fmt.Sprintf("incorrect logger level configuration %s (%v), "+
			"value should be one of %v", lvlStr, err, [...]zapcore.Level{
			zapcore.DebugLevel,
			zapcore.InfoLevel,
			zapcore.WarnLevel,
			zapcore.ErrorLevel,
			zapcore.DPanicLevel,
			zapcore.PanicLevel,
			zapcore.FatalLevel,
		}))
	}

	encoding := v.GetString(cfgLoggerEncoding)
	if encoding != "console" && encoding != "json" {
		panic(fmt.Errorf("invalid logger encoding: %s", encoding))
	}

	c := zap.NewProductionConfig()
	c.Level = zap.NewAtomicLevelAt(lvl)
	c.Encoding = encoding
	c.Sampling = nil
	// If the program is run in TTY, the logger adds a timestamp to its entries.
	if (term.IsTerminal(int(os.Stdout.Fd())) && !v.GetBool(cfgLoggerTimestamp)) || v.GetBool(cfgLoggerTimestamp) {
		c.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
	} else {
		c.EncoderConfig.EncodeTime = func(_ time.Time, _ zapcore.PrimitiveArrayEncoder) {}
	}

	l, err := c.Build(
		zap.AddStacktrace(zap.NewAtomicLevelAt(zapcore.FatalLevel)),
	)
	if err != nil {
		panic(fmt.Sprintf("build zap logger instance: %v", err))
	}

	return l
}

// ServerConfig contains parsed config for the Echo server.
type ServerConfig struct {
	CleanupTimeout  time.Duration
	GracefulTimeout time.Duration
	MaxHeaderSize   int
	ListenLimit     int
	Endpoints       []EndpointInfo
}

const (
	FlagCleanupTimeout  = "cleanup-timeout"
	FlagGracefulTimeout = "graceful-timeout"
	FlagMaxHeaderSize   = "max-header-size"
	FlagListenLimit     = "listen-limit"
)

func BindDefaultFlags(flagSet *pflag.FlagSet) {
	flagSet.Duration(FlagCleanupTimeout, 10*time.Second, "grace period for which to wait before killing idle connections")
	flagSet.Duration(FlagGracefulTimeout, 15*time.Second, "grace period for which to wait before shutting down the server")
	flagSet.Int(FlagMaxHeaderSize, 1000000, "controls the maximum number of bytes the server will read parsing the request header's keys and values, including the request line. It does not limit the size of the request body")
	flagSet.Int(FlagListenLimit, 0, "limit the number of outstanding requests")
}

func serverConfig(v *viper.Viper) *ServerConfig {
	return &ServerConfig{
		CleanupTimeout:  v.GetDuration(cfgServerSection + FlagCleanupTimeout),
		GracefulTimeout: v.GetDuration(cfgServerSection + FlagGracefulTimeout),
		MaxHeaderSize:   v.GetInt(cfgServerSection + FlagMaxHeaderSize),
		ListenLimit:     v.GetInt(cfgServerSection + FlagListenLimit),
		Endpoints:       fetchEndpoints(v),
	}
}

func fetchEndpoints(v *viper.Viper) []EndpointInfo {
	var servers []EndpointInfo

	if v.IsSet(cmdListenAddress) {
		key := cfgServerEndpoints + ".0."
		// If this address is set, we don't use config file to set other parameters.
		serverInfo := EndpointInfo{
			Address:         v.GetString(key + cfgEndpointAddress),
			ExternalAddress: v.GetString(key + cfgEndpointExternalAddress),
			KeepAlive:       v.GetDuration(key + cfgEndpointKeepAlive),
			ReadTimeout:     v.GetDuration(key + cfgEndpointReadTimeout),
			WriteTimeout:    v.GetDuration(key + cfgEndpointWriteTimeout),
		}
		keyFile := v.GetString(key + cfgTLSKeyFile)
		certFile := v.GetString(key + cfgTLSCertFile)
		if keyFile != "" && certFile != "" {
			// If TLS key and certificate are set in the command arguments, we enable TLS.
			serverInfo.TLS.Enabled = true
			serverInfo.TLS.KeyFile = keyFile
			serverInfo.TLS.CertFile = certFile
		}
		servers = append(servers, serverInfo)
	} else {
		for i := 0; ; i++ {
			key := cfgServerEndpoints + "." + strconv.Itoa(i) + "."

			var serverInfo EndpointInfo
			serverInfo.Address = v.GetString(key + cfgEndpointAddress)
			if serverInfo.Address == "" {
				break
			}
			serverInfo.ExternalAddress = v.GetString(key + cfgEndpointExternalAddress)
			serverInfo.KeepAlive = v.GetDuration(key + cfgEndpointKeepAlive)
			serverInfo.ReadTimeout = v.GetDuration(key + cfgEndpointReadTimeout)
			serverInfo.WriteTimeout = v.GetDuration(key + cfgEndpointWriteTimeout)
			serverInfo.TLS.Enabled = v.GetBool(key + cfgTLSEnabled)
			serverInfo.TLS.KeyFile = v.GetString(key + cfgTLSKeyFile)
			serverInfo.TLS.CertFile = v.GetString(key + cfgTLSCertFile)
			serverInfo.TLS.CertCAFile = v.GetString(key + cfgTLSCertCAFile)

			servers = append(servers, serverInfo)
		}
	}

	return servers
}

func newNeofsAPI(ctx context.Context, logger *zap.Logger, v *viper.Viper) (*handlers.RestAPI, error) {
	key, err := getNeoFSKey(logger, v)
	if err != nil {
		return nil, err
	}

	var prm pool.InitParameters
	prm.SetSigner(user.NewAutoIDSignerRFC6979(key.PrivateKey))
	prm.SetNodeDialTimeout(v.GetDuration(cfgNodeDialTimeout))
	prm.SetHealthcheckTimeout(v.GetDuration(cfgHealthcheckTimeout))
	prm.SetClientRebalanceInterval(v.GetDuration(cfgRebalance))
	prm.SetErrorThreshold(v.GetUint32(cfgPoolErrorThreshold))

	for _, peer := range fetchPeers(logger, v) {
		prm.AddNode(peer)
	}

	prometheusConfig := metrics.Config{Enabled: v.GetBool(cfgPrometheusEnabled), Address: v.GetString(cfgPrometheusAddress)}
	if prometheusConfig.Enabled {
		poolStat := metrics.NewPoolMetrics()
		prm.SetStatisticCallback(poolStat.OperationCallback)
	}

	p, err := pool.NewPool(prm)
	if err != nil {
		return nil, err
	}

	if err = p.Dial(ctx); err != nil {
		return nil, err
	}

	ni, err := p.NetworkInfo(ctx, client.PrmNetworkInfo{})
	if err != nil {
		return nil, fmt.Errorf("networkInfo: %w", err)
	}

	var apiPrm handlers.PrmAPI
	apiPrm.Pool = p
	apiPrm.Key = key
	apiPrm.Logger = logger
	apiPrm.NNSName = v.GetString(cfgServerNNSName)

	pprofConfig := metrics.Config{Enabled: v.GetBool(cfgPprofEnabled), Address: v.GetString(cfgPprofAddress)}
	apiPrm.PprofService = metrics.NewPprofService(logger, pprofConfig)
	apiPrm.PrometheusService = metrics.NewPrometheusService(logger, prometheusConfig)
	if prometheusConfig.Enabled {
		apiPrm.GateMetric = metrics.NewGateMetrics()
		apiPrm.GateMetric.SetGWVersion(Version)

		apiPrm.ApiMetric = metrics.NewApiMetrics()
	}

	apiPrm.ServiceShutdownTimeout = defaultShutdownTimeout
	if apiPrm.MaxPayloadBufferSize = ni.MaxObjectSize(); apiPrm.MaxPayloadBufferSize == 0 {
		// default to some heuristic value that in practice should not be needed at all
		apiPrm.MaxPayloadBufferSize = defaultMaxObjectPayloadBufferSize
		logger.Debug("NeoFS max object size setting is zero, using default limit for payload buffer size",
			zap.Uint64("value", apiPrm.MaxPayloadBufferSize))
	}

	apiPrm.DefaultTimestamp = v.GetBool(cfgPoolDefaultTimestamp)

	return handlers.NewAPI(&apiPrm)
}

func fetchPeers(l *zap.Logger, v *viper.Viper) []pool.NodeParam {
	var nodes []pool.NodeParam
	for i := 0; ; i++ {
		key := cfgPeers + "." + strconv.Itoa(i) + "."
		address := v.GetString(key + cfgPeerAddress)
		weight := v.GetFloat64(key + cfgPeerWeight)
		priority := v.GetInt(key + cfgPeerPriority)

		if address == "" {
			break
		}
		if weight <= 0 { // unspecified or wrong
			weight = 1
		}
		if priority <= 0 { // unspecified or wrong
			priority = 1
		}

		nodes = append(nodes, pool.NewNodeParam(priority, address, weight))

		l.Info("added connection peer",
			zap.String("address", address),
			zap.Int("priority", priority),
			zap.Float64("weight", weight),
		)
	}

	return nodes
}
